src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2376,45 +2372,131 @@ private void SelectVoice(string voice)
+                            Log.InfoFormat("StepListCommand changing keyboard to string {0}. Replace previous keyboard={1}", commandKeyValue.String, vReturn);
                         }
+                        KeySelectionResult(commandKeyValue, multiKeySelection);
                     }
-                    else if (vStep.StartsWith("Text>"))
+                    else if (commandList.First().StartsWith("KeyToggle>"))
                     {
-                        keyboardOutputService.ProcessSingleKeyText(vStep.Substring(5));
+                        commandKeyValue = new KeyValue(commandList.First().Substring(10));
+                        Log.InfoFormat("StepListCommand: key toggle on {0} key", commandKeyValue.String);
+
+                        if (keyStateService.KeyDownStates[singleKeyValue].Value != KeyDownStates.Up)
+                        {
+                            Log.Info("StepListCommand: ...toggling key up");
+                            await keyboardOutputService.ProcessSingleKeyPress(commandKeyValue.String, KeyPressKeyValue.KeyPressType.Release);
+                            //singleKeyValue represents the visual item
+                            keyStateService.KeyDownStates[singleKeyValue].Value = KeyDownStates.Up;
+                            //commandKeyValue represents the virtual key character
+                            keyStateService.KeyDownStates[commandKeyValue].Value = KeyDownStates.Up;
+                        }
+                        else
+                        {
+                            Log.Info("StepListCommand: ...toggling key down");
+                            await keyboardOutputService.ProcessSingleKeyPress(commandKeyValue.String, KeyPressKeyValue.KeyPressType.Press);
+                            //singleKeyValue represents the visual item
+                            keyStateService.KeyDownStates[singleKeyValue].Value = KeyDownStates.Down;
+                            //commandKeyValue represents the virtual key character
+                            keyStateService.KeyDownStates[commandKeyValue].Value = KeyDownStates.Down;
+                        }
                     }
-                    else if (vStep.StartsWith("Wait>"))
+                    else if (commandList.First().StartsWith("KeyDown>"))
                     {
-                        System.Threading.Thread.Sleep(int.Parse(vStep.Substring(5)));
+                        commandKeyValue = new KeyValue(commandList.First().Substring(8));
+                        Log.InfoFormat("StepListCommand: key down on {0} key", commandKeyValue.String);
+                        await keyboardOutputService.ProcessSingleKeyPress(commandKeyValue.String, KeyPressKeyValue.KeyPressType.Press);
+                        //singleKeyValue represents the visual item
+                        keyStateService.KeyDownStates[singleKeyValue].Value = KeyDownStates.Down;

```

Comment:
Is there any danger of the singleKeyValue being null? Can you pls give an example of what the singleKeyValue might be compared to the commandKeyValue so I can understand? Could you also pls add these examples to the comments here?

Reply:
SingleKeyValue is the full array of action/value commands associated with a Dynamic key and looks like this:
`<StepList><Loop>10<KeyDown>2<Wait>1000<KeyUp>2</Loop>`

CommandKeyValue is what you get when you pull out one of the values and would look like this:
`2`

Reply:
@AdamRoden Understood. Would you please add a comment to that effect early in the method so people coming to the code fresh can understand easily?